### ISSUE TITLE
OOTB not working due to CredentialsFilter

### ISSUE THREAD

**Issue by twocs** [OSLC/iotp-adaptor#6]
I followed the steps in the docs but the iotp fails to launch.

> 29-Jan-2020 10:07:52.030 SEVERE [main] org.apache.catalina.core.StandardContext.filterStart Exception starting filter [CredentialsFilter]
> ...
> Caused by: java.lang.NullPointerException: inStream parameter is null
> 		at java.base/java.util.Objects.requireNonNull(Objects.java:247)
> 		at java.base/java.util.Properties.load(Properties.java:404)
> 		at com.ibm.oslc.adaptor.iotp.impl.CredentialsFilter.<clinit>(CredentialsFilter.java:85)

This line of the [CredentialsFilter.java](https://github.com/OSLC/iotp-adaptor/blob/master/iotp-adaptor/src/main/java/com/ibm/oslc/adaptor/iotp/impl/CredentialsFilter.java) code tries to load /config.properties.
`85			configProperties.load(TRSObject.class.getResourceAsStream("/config.properties"));
`

Is config.properties a file? This file does not exist. Is that because it's in a [.gitignore](https://github.com/OSLC/iotp-adaptor/blob/master/iotp-adaptor/src/main/resources/.gitignore)? The instructions do not mention this file. It is referred to in multiple parts of the code. Only two other lyo projects on GitHub have a config.properties file ([org.eclipse.lyo.adapter.magicdraw/configuration/config.properties)](https://github.com/eclipse/lyo.adapter-magicdraw/blob/master/org.eclipse.lyo.adapter.magicdraw/configuration/config.properties), [oslc4simulink/configuration/config.properties](https://github.com/ld4mbse/oslc-adapter-simulink/blob/master/edu.gatech.mbsec.adapter.simulink/oslc4jsimulink%20configuration/config.properties)) but I have no information about where this file should go and what should be in this file, if that is really the problem.

I want to get iotp-adaptor running, but what step am I missing?

**Comment by jadelkhoury** [comment]
Hi

i assume you are following the instructions for this code from https://oslc.github.io/developing-oslc-applications/iotp-adaptor-sample.html ? 

config.properties is a file, and it does seem to be missing. Normally, such files would contain the configurations of a specific installation, and hence not normally committed. but it would have been nice with an "example" file to know what to include.

But it should not be hard to work out what to include.
Try to work what what the code is trying to do with this "configProperties" variable. You will see that it is trying to read properties (trs_user, etc. See for example line 179)
Then create your file, with lines like (simply property_name=propertyvalues)
tsr_user=xxx

The file should be placed under iotp-adaptor/iotp-adaptor/src/main/resources/

Hope this helps. I have not run the code myself.
Jad

**Comment by twocs** [comment]
Thanks @jadelkhoury. I'm following these instructions from a couple of clicks past the link you suggested: https://oslc.github.io/developing-oslc-applications/iotp_adaptor/userGuide/install-and-config.html.

Adding this file got past the first hurdle:

iotp-adaptor/iotp-adaptor/src/main/resources/config.properties
```
trs_user=<user>
trs_user_password=<password>
```

The next error seems to be that the system is trying to connect to Bluemix (Watson IoT Platform resources?), possibly using the username and password from config.properties.

**Comment by jadelkhoury** [comment]
Hi tom

And the error is?

**Comment by berezovskyi** [comment]
@twocs do you have an account with Watson IoT Platform?

**Comment by twocs** [comment]
Thanks for your messages. I'm slowly working through it. 

Specifically the error is that there's no more Bluemix since 2018, so the urls need updating. I've stripped out the Bluemix references. Would be interesting to know the updated urls now that it's been renamed IBM Cloud. The docs were not very clear about how

At the moment I'm most interested in setting up the OAuth server for a friend relationship in DNG.

There's also another file expected (https://github.com/OSLC/lyo-samples/blob/master/oauth-cli-helper/src/main/java/org/eclipse/lyo/client/oauth/sample/OAuthClient.java):

src/main/resources/oauth.properties
```
consumer_key=<consumer_key>
consumer_secret=<consumer_secret>
```

**Comment by jadelkhoury** [comment]
Are you trying to setup an OAuth server or client? From the file you share, it seems to be a client. But you said server.

if you are a client to an oauth server, you need to (manually) get a consumer_key & consumer_secret that can be used to identify your client. Think of it as a username/password, but for your client application, and not for a human.
How to get them? The oauth should have a GUI that lets you create/register your client, and hence lets you create a consumer key/secret.

**Comment by twocs** [comment]
@jadelkhoury The client is (only) necessary for the tests in iotp-adaptor/src/test/java/com/ibm/oslc/adaptor/tests/CE4IoTConnectorTests.java

**Comment by twocs** [comment]
> Are you trying to setup an OAuth server or client? 

I need to setup an OSLC server to be recognised by Doors NG as a friend. This means an OAuth server is required. This is only project that I've found online that suggests it can be used this way, but unfortunately it seems that some urls are not up-to-date.

> @twocs do you have an account with Watson IoT Platform?

The first thing I tried was to get a free Watson IoT Platform plan. But there are many hard-coded references to Bluemix. I might try @isccarrasco branch that seems to address the urls issue (and authentication with IBM Cloud). https://github.com/isccarrasco/iotp-adaptor

**Comment by jamsden** [comment]
I don't know if iotp-adaptor works with the new IBM Cloud. It is possible that the URLs have changed, accounts are no longer active, authentication has been updated, keys have expired, etc.

**Comment by jadelkhoury** [comment]
@twocs If you just need something to work with Doors NG, maybe your better bet would be to look at the original OSLC Workshop that implements an implements an adaptor to Bugzilla.
You can specifically look at Lab 6 (page 43) for some oauth instructions to connect to RTC.

https://github.com/eclipse/lyo.docs/blob/master/lyo-rest-workshop/Lyo_OSLC_Workshop.pdf

**Comment by twocs** [comment]
@jamsden @jadelkhoury Thanks, I will take a closer look at the Bugzilla.

After the logins to the cloud platform and swapping over to @isccarrasco's fork, which contains the new urls and authentications, I have been able to get the OSLC/IOTP-Adaptor to connect to IBM Cloud, and got DoorsNG to accept it as a friend. However, I'm currently stuck on adding the association. The dropdown list isn't populating as expected. I hope the Bugzilla lab is helpful in that regard.

**Comment by berezovskyi** [comment]
@isccarrasco should we merge your changes in?

**Comment by jamsden** [comment]
If adding an association doesn't show the IoT Platform server, then the rootservices probably isn't correct. It should provide rmServiceProviders and cmServiceProviders corresponding to the IoT Platform. If you see the server, but no organizations, then perhaps your login didn't work, or there are no organizations for the user you used. The service providers are constructed dynamically from the organizations the user is a member of.

**Comment by jamsden** [comment]
I just noticed that  @isccarrasco's fork uses apikey instead of grant_type password. That may mean that after login, there is no user established for which to determine organization membership. So getting organizations for the user may be empty.

**Comment by jadelkhoury** [comment]
Can you share the content of the rootservices file you are producing? 
One common problem I normally face is that the URLs in the rootservices are not exactly the same as those of the server, due to problems in the configuration of the server. For example, the rootservices URL is "http://localhost....", while the serviceProviders URLs are pointing to "http:\\127.0,0.1". Some applications don't like that.

**Comment by isccarrasco** [comment]
Hello guys!

I'm not sure whether IBM Cloud has the grant_type password right now or it still missing, when I was working on this fork, I tried to use the password grant_type but didn't find how to implement it, that because I changed to the apiKey.

I'll review this fork this week for testing it again, maybe then I could resend the PR.

Regards!

**Comment by twocs** [comment]
> Can you share the content of the rootservices file you are producing?
> One common problem I normally face is that the URLs in the rootservices are not exactly the same as those of the server, due to problems in the configuration of the server. For example, the rootservices URL is "http://localhost....", while the serviceProviders URLs are pointing to "http:\127.0,0.1". Some applications don't like that.

I found that there are not a lot of example OSLC files online. I'd think it could be useful. This is the one it's using (I anonymised the URLs).

```
<?xml version="1.0" encoding="UTF-8"?>
<rdf:Description rdf:about="https://example.com:8443/rootservices"
    xmlns:iotp="http://jazz.net/ns/dm/iotp#"
    xmlns:oslc="http://open-services.net/ns/core#"
    xmlns:oslc_am="http://open-services.net/ns/am#"
    xmlns:oslc_rm="http://open-services.net/xmlns/rm/1.0/"
    xmlns:oslc_cm="http://open-services.net/xmlns/cm/1.0/"
    xmlns:trs="http://open-services.net/ns/core/trs#"
    xmlns:dc="http://purl.org/dc/terms/"
    xmlns:jfs="http://jazz.net/xmlns/prod/jazz/jfs/1.0/" 
	xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">

	<dc:title>OSLC AM CE4IoTConnector Jazz Root Services</dc:title>

	<!-- Service Providers - ServiceProviderCatalog URLs for each domain -->
    <oslc_am:amServiceProviders rdf:resource="https://example.com:8443/services/catalog/singleton" />
	<oslc_rm:rmServiceProviders rdf:resource="https://example.com:8443/services/catalog/singleton" />
	<oslc_cm:cmServiceProviders rdf:resource="https://example.com:8443/services/catalog/singleton" />
	
	
	<!-- <oslc_am:amServiceProviders rdf:resource="https://example.com/am/serviceProviderCatalog"/> -->

    <!-- OAuth URLs for establishing server-to-server connections -->
	<jfs:oauthRealmName>IoTPlatform</jfs:oauthRealmName>
	<jfs:oauthDomain>https://example.com:8443</jfs:oauthDomain>
	<jfs:oauthRequestConsumerKeyUrl rdf:resource="https://example.com:8443/services/oauth/requestKey" />
	<jfs:oauthApprovalModuleUrl rdf:resource="https://example.com:8443/services/oauth/approveKey" />
	<jfs:oauthRequestTokenUrl rdf:resource="https://example.com:8443/services/oauth/requestToken"/>
	<jfs:oauthUserAuthorizationUrl rdf:resource="https://example.com:8443/services/oauth/authorize" />
	<jfs:oauthAccessTokenUrl rdf:resource="https://example.com:8443/services/oauth/accessToken"/>

	<!-- IoT Platform Tracked Resource Set Provider -->
	<iotp:TrackedResourceSetProvider>
		<trs:TrackedResourceSet>
			<trs:trackedResourceSet rdf:resource="https://example.com:8443/services/trs" />
	        <dc:title>iotp-adaptor TRS resources</dc:title>
	        <dc:description>TRS 2.0 provider for IoT Platform resources</dc:description>
	        <dc:type     rdf:resource="http://open-services.net/ns/cm#" />
	        <oslc:domain rdf:resource="http://open-services.net/ns/rm#" />
	        <oslc:domain rdf:resource="http://open-services.net/ns/am#" />
        </trs:TrackedResourceSet>
	</iotp:TrackedResourceSetProvider>		
</rdf:Description>

```

**Comment by twocs** [comment]
I feel like I'm getting closer and closer. DoorsNG was finally putting something in the box:
"No artifact container catalogs available in the selected application for the possible associations."
![Annotation 2020-02-14 134702](https://user-images.githubusercontent.com/72763/74498622-1e62ba80-4f31-11ea-9b5b-0f4ed083c739.png)

I wasn't not clear if that's because there are not the expected devices listed in the catalog, the catalogs urls are simply not setup how DoorsNG likes them. I am able to see the test device when I manually navigate through the tree
rootservices ->
https://example.com:8443/services/catalog/singleton ->
https://example.com:8443/services/iotp/knte64 ->
https://example.com:8443/services/iotp/knte64/resources/selector

So then I created simulated devices in the Watson IotP, and assigned them some events. Apparently the dummy device that I had created earlier wasn't being detected, and the simulated devices are (but only when logged into the page). Not sure exactly what I did, but it seems to be working. When we have an API key, we can use "apikey" for the username and the actual key for the password.

![image](https://user-images.githubusercontent.com/72763/74499567-0d677880-4f34-11ea-967a-5eef36899fd9.png)

Now in DoorsNG I can actually add associations. So I will say that @isccarrasco's fork is very close to ready, but there may be some things that need to be done to get it functional (such as having things in IBM Cloud to be seen). I did remove the duplicate cm and set 

I seem to remember a related document on setting up Node Red for IOTP-Adaptor, but can't find it. Would be good to know the steps to set up IBM Cloud iotp and Node Red for this project.

**Comment by jadelkhoury** [comment]
Hi @twocs 
Just wondering if you made any progress to make the IOTP adaptor to work. Any changes to the current code base that ought to be fixed?

**Comment by twocs** [comment]
I did get it to work by commenting out the Bluemix parts to focus on the IOTP devices section, and then using @isccarrasco's fork with the new IBM method of login, and also activating the simulated IoT in IBM Cloud. The username was "apikey" and the password was the key.

Finally I was able to generate the code I needed using Lyo designer after studying the iotp-adaptor repo the Bugzilla one, and also different examples at https://oslc.github.io/developing-oslc-applications/sample-applications-and-code.html.

As for this iotp-adaptor git repository, the main comment I'd say is that it is out of date because IBM has updated the urls and the login methods. The fork by @isccarrasco nearly completes all that.

It's interesting that the OSLC examples outlive their services: the stock market sample that I started with relies on a now-defunct Google stock API. OSLC doesn't address the logins or services, but it's easier to see how an example works when logins and services still work.
